### PR TITLE
Added/fixed sigma RPCs

### PR DESCRIPTION
**Pull request by martun**

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -3632,12 +3691,12 @@ UniValue listspendzerocoins(const UniValue &params, bool fHelp) {
...
 
     UniValue ret(UniValue::VARR);
-    const CWallet::TxItems & txOrdered = pwalletMain->wtxOrdered;
+    const CWallet::TxItems& txOrdered = pwalletMain->wtxOrdered;
 
     for (CWallet::TxItems::const_reverse_iterator it = txOrdered.rbegin(); it != txOrdered.rend(); ++it) {
         CWalletTx *const pwtx = (*it).second.first;
 
-        if (!pwtx || !pwtx->IsZerocoinSpend() || pwtx->vin.size() != 1)
+        if (!pwtx || !pwtx->IsZerocoinSpendV3() || pwtx->vin.size() != 1)
```
Comment (thebevrishot) [inline src/wallet/rpcwallet.cpp]: sigma spend allow vin more than 1.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -3651,22 +3710,20 @@ UniValue listspendzerocoins(const UniValue &params, bool fHelp) {
         entry.push_back(Pair("abandoned", pwtx->isAbandoned()));
 
         const CTxIn &txin = pwtx->vin[0];
-        int pubcoinId = txin.nSequence;
-        bool fModulusV2 = pubcoinId >= ZC_MODULUS_V2_BASE_ID;
-        if (fModulusV2)
-            pubcoinId -= ZC_MODULUS_V2_BASE_ID;
+        // For sigma public coin group id is prevout.n.
```
Comment (thebevrishot) [inline src/wallet/rpcwallet.cpp]: should fix here to for-each vins too.

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -3436,33 +3492,39 @@ UniValue listpubcoins(const UniValue& params, bool fHelp) {
...
+            "\nArguments:\n"
+            "1. <all> (string, optional) 0.1,0.5,1,10,25,100 (default) to return all sigma public coins with given denomination. empty to return all pubcoin.\n"
+            "\nResults are an array of Objects, each of which has:\n"
+            "{id, IsUsed, denomination, value, serialNumber, nHeight, randomness}";
+    if (fHelp || params.size() > 1) {
+        throw runtime_error(help_message);
+    }
 
-    int64_t denomination = -1;
+    CoinDenominationV3 denomination;
```
Comment (nopslide) [inline src/wallet/rpcwallet.cpp]: CoinDenominationV3 ??

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -3436,33 +3492,39 @@ UniValue listpubcoins(const UniValue& params, bool fHelp) {
...
     if (params.size() > 0) {
-        denomination = params[0].get_int();
+        filter_by_denom = true;
+        if (!sigma::StringToDenomination(params[0].get_str(), denomination)) {
+            throw runtime_error(help_message);
+        }
     }
 
-    list <CZerocoinEntryV3> listPubcoin;
+    list<CZerocoinEntryV3> listPubcoin;
```
Comment (nopslide) [inline src/wallet/rpcwallet.cpp]: though it changed to CSigmaEntry

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -3436,33 +3492,39 @@ UniValue listpubcoins(const UniValue& params, bool fHelp) {
...
     CWalletDB walletdb(pwalletMain->strWalletFile);
     walletdb.ListPubCoinV3(listPubcoin);
     UniValue results(UniValue::VARR);
-    listPubcoin.sort(CompHeightV3);
+    listPubcoin.sort(CompIDV3);
```
Comment (nopslide) [inline src/wallet/rpcwallet.cpp]: V3 :)

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -3651,22 +3710,20 @@ UniValue listspendzerocoins(const UniValue &params, bool fHelp) {
...
+        // OP_ZEROCOINSPENDV3 is written. In zerocoin you will see +4 instead,
+        // because the size of serialized spend is also written, probably in 3 bytes.
+        CDataStream serializedCoinSpend((const char *)&*(txin.scriptSig.begin() + 1),
                                         (const char *)&*txin.scriptSig.end(),
                                         SER_NETWORK, PROTOCOL_VERSION);
-        libzerocoin::CoinSpend spend(fModulusV2 ? ZCParamsV2 : ZCParams, serializedCoinSpend);
-        int spendVersion = spend.getVersion();
-
-        entry.push_back(Pair("denomination", (int)spend.getDenomination()));
+        sigma::ParamsV3* zcParams = sigma::ParamsV3::get_default();
```
Comment (nopslide) [inline src/wallet/rpcwallet.cpp]: sigma::ParamsV3 or sigma::Params ?

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -3711,7 +3768,7 @@ UniValue listspendzerocoinsV3(const UniValue &params, bool fHelp) {
     for (CWallet::TxItems::const_reverse_iterator it = txOrdered.rbegin(); it != txOrdered.rend(); ++it) {
         CWalletTx *const pwtx = (*it).second.first;
 
-        if (!pwtx || !pwtx->IsZerocoinSpendV3() || pwtx->vin.size() != 1)
+        if (!pwtx || !pwtx->IsZerocoinSpend() || pwtx->vin.size() != 1)
```
Comment (nopslide) [inline src/wallet/rpcwallet.cpp]: I think this checking logic may be incorrect.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -2829,6 +2824,57 @@ void CWallet::ListAvailableCoinsMintCoins(vector <COutput> &vCoins, bool fOnlyCo
     }
 }
 
+void CWallet::ListAvailableSigmaMintCoins(vector<COutput> &vCoins, bool fOnlyConfirmed) const {
+    vCoins.clear();
+    LOCK(cs_wallet);
+    list<CZerocoinEntryV3> listOwnCoins = list<CZerocoinEntryV3>();
```
Comment (nopslide) [inline src/wallet/wallet.cpp]: V3 entry

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (martun) [reply src/wallet/rpcwallet.cpp]: thanks

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (martun) [reply src/wallet/rpcwallet.cpp]: fixed

**Pull request merged**